### PR TITLE
Add pure functional dropout

### PR DESCRIPTION
**Pull request by junjihashimoto**
This PR provides pure functional dropout.
dropout implementation is here.
When training, it uses _fused_dropout. _fused_dropout can use local generater.
https://github.com/pytorch/pytorch/blob/1.6/aten/src/ATen/native/Dropout.cpp
https://github.com/pytorch/pytorch/blob/1.6/aten/src/ATen/native/native_functions.yaml#L167

### CODE REVIEW

File: hasktorch/src/Torch/Random.hs [hunk 1]
```diff
@@ -105,3 +131,12 @@ normal mean std size opts gen = generatorFactory (LibTorch.normal_ddlGo (realToF
...
 normal' :: Double -> Double -> [Int] -> Generator -> (Tensor,Generator)
 normal' mean std size gen = normal mean std size defaultOpts gen
+
+-- | Pure functional dropout
+dropout
+  :: Double -- ^ dropout probability
+  -> Maybe Generator -- ^ generator
+  -> Tensor -- ^ input
+  -> (Tensor,Generator) -- ^ output
+dropout p Nothing input = cast3 ATen.dropout_tdb input p False
```
Comment (tscholak) [inline hasktorch/src/Torch/Random.hs:141]: Hi @junjihashimoto, this looks great, thank you for working on this.
This line here is odd, though. I don't understand how this can return a `Generator` when `Nothing` is passed into it. Isn't `ATen.dropout_tdb` the one that returns `IO Tensor`?

Comment (junjihashimoto) [reply hasktorch/src/Torch/Random.hs:141]: Thank you for your pointing it out. If the input generator is Nothing, the output generator should also return Nothing or just a Tensor(not the Tuple).
